### PR TITLE
init algorithm

### PR DESCRIPTION
**Pull request by christianbean**
This moves some of the code in the initialiser from Tiling to an algorithm class

### CODE REVIEW

File: tilings/algorithms/gridded_perm_reduction.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    def _minimize(self) -> None:
+        """
+        Removes non minimal gridded permutations from the set.
+        """
+        res: List[GriddedPerm] = []
+        for gp_to_add in self.griddedperms:
+            if all(gp_already_added not in gp_to_add for gp_already_added in res):
+                res.append(gp_to_add)
+        self.griddedperms = tuple(sorted(res))
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:23]: We can think about how to optimize this.

File: tilings/algorithms/gridded_perm_reduction.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        The set of obstructions are first reduced to a minimal set. The
+        requirements that contain any obstructions are removed from their
+        respective lists. If any requirement list is empty, then the tiling is
+        empty.
+        """
+        while True:
+            # Minimize the set of obstructions
+            minimized_obs = MinimalGriddedPermSet(
+                [self._clean_isolated(ob) for ob in self._obstructions]
+            )
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:102]: bad

File: tilings/algorithms/gridded_perm_reduction.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                    for f in gp.factors():
+                        # if factor implied by some requirement list then we
+                        # remove it from the gridded perm
+                        if not any(
+                            all(f in g for g in other_requirement)
+                            for other_requirement in requirements
+                            if not all(
+                                any(g2 in g1 for g2 in requirement)
+                                for g1 in other_requirement
+                            )
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:185]: Hmmmm

File: tilings/algorithms/gridded_perm_reduction.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            if (
+                self._obstructions == minimized_obs
+                and self._requirements == minimized_requirements
+            ):
+                break
+
+            self._obstructions = minimized_obs
+            self._requirements = minimized_requirements
+        GriddedPermReduction.passes[i] += 1
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:85]: # if there is a requirement for which every component contains the same factor of
    # obstruction, then that factor can be removed from obstruction

File: tilings/algorithms/gridded_perm_reduction.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+                    for factor in factors
+                ):
+                    idx_to_remove.add(i)
+                    break
+        return sorted(
+            requirement
+            for idx, requirement in enumerate(requirements)
+            if idx not in idx_to_remove
+        )
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:261]: # detect if gridded_perm is contained in every gp of requirement

File: tilings/algorithms/gridded_perm_reduction.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    @cssmethodtimer("GriddedPermReduction._griddedperm_implied_by_requirement")
+    def _griddedperm_implied_by_requirement(
+        self, griddedperm: GriddedPerm, requirement: Iterable[GriddedPerm]
+    ):
+        """
+        Return True, if the containment of the gridded perm is implied by
+        the containment of the requirement.
+        """
+        return all(griddedperm in gp for gp in requirement)
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:271]: # detect if there exists a requirement such that every component in that requirement
    # contains griddedperm

File: tilings/algorithms/gridded_perm_reduction.py (hunk 4 above)
Comment (jaypantone) [reply tilings/algorithms/gridded_perm_reduction.py:85]: # uses requirements to shrink obstructions

File: tilings/algorithms/gridded_perm_reduction.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+            partial(self.remove_avoided, obstructions=obstructions),
+            self.remove_redundant_requirements,
+        )
+        minimized_requirements = self._requirements
+        for algo in algos:
+            minimized_requirements = algo(minimized_requirements)
+            if any(not r for r in minimized_requirements):
+                return (tuple(),)
+        return tuple(minimized_requirements)
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:115]:
```
# If all of the gp's in a requirment list contain the same factor, you can remove
# that factor from all of them and add it as it's own size one requirement list
```

File: tilings/algorithms/gridded_perm_reduction.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+                for factor in factors:
+                    res.append((factor,))
+                res.append(
+                    tuple(
+                        gp.get_gridded_perm_in_cells(remaining_cells)
+                        for gp in requirement
+                    )
+                )
+        return res
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:154]:
```
    # if a gridded perm G in a requirment has a factor that appears in EVERY gridded
    # perm of some other requirement, then that factor can be removed from G
```

File: tilings/algorithms/gridded_perm_reduction.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+                                other_requirement, requirement
+                            )
+                        ),
+                    ):
+                        cells.extend(f.pos)
+                newgp = gp.get_gridded_perm_in_cells(cells)
+                newgps.append(newgp)
+            cleaned_reqs.append(tuple(newgps))
+        return cleaned_reqs
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:188]:
```# We can remove any gridded perm in any requiremenet if the gridded perm contains
# an obstruction
```

File: tilings/algorithms/gridded_perm_reduction.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+            cleanreq = tuple(
+                gp for gp in self._minimize(requirement) if gp.avoids(*obstructions)
+            )
+            # If cleanreq is empty, then can not contain this requirement so
+            # the tiling is empty.
+            if not cleanreq:
+                return [tuple()]
+            res.append(cleanreq)
+        return res
+
```
Comment (jaypantone) [inline tilings/algorithms/gridded_perm_reduction.py:212]:
```
    # If a requirement contains a gridded perm, each of whose factors is implied by
    # a (possibly different) requirement, then this gridded perm always exists and so
    # the whole requirement can be deleted.
```

File: tilings/algorithms/gridded_perm_reduction.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+            if not all(requirement):
+                continue
+            newgps: List[GriddedPerm] = []
+            for gp in requirement:
+                cells: List[Cell] = []
+                for f in gp.factors():
+                    if not self._griddedperm_implied_by_some_requirement(
+                        f,
+                        (
+                            other_requirement
```
Comment (enadeau) [inline tilings/algorithms/gridded_perm_reduction.py:204]: Can we just replace this big thing by
`requirements[:idx] + requirements[idx+1:]`

Comment (enadeau) [reply tilings/algorithms/gridded_perm_reduction.py:204]: Or if your concern about creating new list we can also use itertools.islice

File: tilings/tiling.py [hunk 12]
```diff
@@ -64,16 +67,31 @@ class Tiling(CombinatorialClass):
...
+        remove_empty_rows_and_cols: bool = True,
         derive_empty: bool = True,
-        minimize: bool = True,
+        simplify: bool = True,
         sorted_input: bool = False,
+        already_minimized_obs: bool = False,
     ) -> None:
+        """
+        - if not sorted_input, input will be sorted
+        - if minimize, will be minimized
```
Comment (enadeau) [inline tilings/tiling.py:87]: This doctrstring does not match the actual arguments of the method

File: tilings/tiling.py [hunk 13]
```diff
@@ -125,37 +145,38 @@ def from_perms(
...
     def _fill_empty(self) -> None:
         """Add size one obstructions to cells that are empty."""
         add = []
         (i, j) = self.dimensions
+        active_cells = self.active_cells
+        empty_cells = self.empty_cells
         for x in range(i):
             for y in range(j):
-                if (x, y) not in self.active_cells and (x, y) not in self.empty_cells:
+                if (x, y) not in active_cells and (x, y) not in empty_cells:
```
Comment (enadeau) [inline tilings/tiling.py:157]: I would replace this big nested statement by
```python
all_cells = set(itertools.product(*map(range, dimensions)))
add = {GriddedPerm.single_cell(Perm((0,)), cell) for cell in all_cells - active_cells - empty_cells}
```
but maybe I'm just obsessed about avoiding big nesting
```

File: tilings/tiling.py [hunk 14]
```diff
@@ -928,8 +811,11 @@ def sub_tiling(self, cells: Iterable[Cell], factors: bool = False) -> "Tiling":
             if (factors and ass.gps[0].pos[0] in cells)
             or all(c in cells for c in chain.from_iterable(gp.pos for gp in ass.gps))
         )
-        return self.__class__(obstructions, requirements, assumptions)
+        return self.__class__(
+            obstructions, requirements, assumptions, simplify=False, sorted_input=True
+        )
 
+    @cssmethodtimer("Tiling.find_factors")
```
Comment (enadeau) [inline tilings/tiling.py:818]: Why do we time the `find_factors` methods?

File: tilings/algorithms/gridded_perm_reduction.py [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+            minimized_requirements = self.minimal_reqs(minimized_obs)
+            minimized_requirements = tuple(
+                sorted(tuple(sorted(set(req))) for req in minimized_requirements)
+            )
+
+            if minimized_requirements and not minimized_requirements[0]:
+                set_empty()
+                break
+
+            if self._requirements == minimized_requirements:
```
[the remark this answers is not in the document]
Comment (jaypantone) [reply tilings/algorithms/gridded_perm_reduction.py:81]: No, this is one of the improvements I made. The `minimal_obs` function uses the requirements to simplify obstructions (and is idempotent). So if, in one pass, the obstructions change but the requirements don't, there is still no need to loop again.

[the remark this answers is not in the document]
Comment (jaypantone) [reply tilings/algorithms/gridded_perm_reduction.py:81]: @christianbean Do you agree with this logic?

File: tilings/algorithms/gridded_perm_reduction.py (hunk 11 above)
Comment (jaypantone) [reply tilings/algorithms/gridded_perm_reduction.py:204]: I'm not sure I understand this comment. This function shrinks GPs in requirements by removing some factors.

File: tilings/tiling.py (hunk 13 above)
Comment (jaypantone) [reply tilings/tiling.py:157]: Do you claim your version is faster? I find it definitely less readable, but maybe that's worth it if there's a speed advantage.

File: tilings/tiling.py (hunk 14 above)
Comment (jaypantone) [reply tilings/tiling.py:818]: In case it's slow? We can probably remove this.

File: tilings/algorithms/gridded_perm_reduction.py (hunk 11 above)
Comment (enadeau) [reply tilings/algorithms/gridded_perm_reduction.py:204]: I meant that the second argument of `_griddedperm_implied_by_some_requirement` could be replaced  by this which is much shorted.

File: tilings/tiling.py (hunk 12 above)
Comment (jaypantone) [reply tilings/tiling.py]: Fixed.

File: tilings/tiling.py (hunk 13 above)
Comment (enadeau) [reply tilings/tiling.py:157]: Not claiming that it's faster only that it is more readable. Probably not worth changing if you don't find it more readable

File: tilings/algorithms/gridded_perm_reduction.py (hunk 11 above)
Comment (jaypantone) [reply tilings/algorithms/gridded_perm_reduction.py:204]: Oh, I see. I think itertools.islice would be okay. @christianbean?

File: tilings/tiling.py (hunk 14 above)
Comment (enadeau) [reply tilings/tiling.py:818]: My question came from the fact that it's the only method not related to the initializer that we are timing. We might have added the timing by accident.

Comment (jaypantone) [reply tilings/tiling.py:818]: I think it was used at one point, but maybe not anymore. We should remove it.

File: tilings/algorithms/gridded_perm_reduction.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (enadeau) [reply tilings/algorithms/gridded_perm_reduction.py:81]: Took a closer look. I agree with the logic

[the remark this answers is not in the document]
Comment (christianbean) [reply tilings/algorithms/gridded_perm_reduction.py:81]: I agree too

File: tilings/algorithms/gridded_perm_reduction.py (hunk 11 above)
Comment (christianbean) [reply tilings/algorithms/gridded_perm_reduction.py:204]: That is the best option

File: tilings/tiling.py (hunk 14 above)
Comment (jaypantone) [reply tilings/tiling.py]: Fixed. Added timing to Tiling._simplify_griddedperms.

File: tilings/algorithms/gridded_perm_reduction.py (hunk 11 above)
Comment (jaypantone) [reply tilings/algorithms/gridded_perm_reduction.py]: Fixed.

File: tilings/algorithms/gridded_perm_reduction.py [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+            GriddedPermReduction.factored_reqs,
+            self.cleaned_requirements,
+            partial(self.remove_avoided, obstructions=obstructions),
+        )
+        minimized_requirements = self._requirements
+        for algo in algos:
+            minimized_requirements = algo(minimized_requirements)
+            if any(not r for r in minimized_requirements):
+                return (tuple(),)
+        return tuple(sorted(minimized_requirements))
```
Comment (christianbean) [inline tilings/algorithms/gridded_perm_reduction.py:138]: I think this sort is redundant

File: tilings/algorithms/gridded_perm_reduction.py [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+            newgps: List[GriddedPerm] = []
+            for gp in requirement:
+                cells: List[Cell] = []
+                for f in gp.factors():
+                    if not self._griddedperm_implied_by_some_requirement(
+                        f,
+                        itertools.chain(
+                            itertools.islice(requirements, idx),
+                            itertools.islice(requirements, idx + 1, None),
+                        ),
```
Comment (christianbean) [inline tilings/algorithms/gridded_perm_reduction.py:207]: we should just import chain and islice

File: tilings/tiling.py [hunk 18]
```diff
@@ -35,13 +36,15 @@
...
     Fusion,
+    GriddedPermReduction,
     GriddedPermsOnTiling,
     MinimalGriddedPerms,
     ObstructionTransitivity,
     RequirementPlacement,
     RowColSeparation,
     SubobstructionInferral,
 )
+from .algorithms.gridded_perm_reduction import func_calls, func_times
```
Comment (christianbean) [inline tilings/tiling.py:47]: probably makes more sense for these to be in tiling.py file when fist defined

Comment (christianbean) [reply tilings/tiling.py:47]: this leads to circular imports so ignore this for now

**Pull request merged**